File: algo_basic_python_java/w9.-sorting/problems/464.sort_integers_ii/464_sort_intergers_ii_merge_sort.java
public class Solution {
    /**
     * @param A: an integer array
     * @return: nothing
     */
    public void sortIntegers2(int[] A) {
        // write your code here
        int[] tmp = new int[A.length];
        mergeSort(A, 0, A.length - 1, tmp);
    }
    private void mergeSort(int[] A, int start, int end, int[] tmp) {
        // when split into single element, return
        if (start >= end) {
            return;
        }
        int mid = start + (end - start)/2; 
        // keep split in half
        mergeSort(A, start, mid, tmp);
        mergeSort(A, mid + 1, end, tmp);
        // merge two half array into tmp
        merge(A, start, end, tmp);
    }
    
    private void merge(int[] A, int start, int end, int[] tmp) {
        int n = end - start + 1;
        int mid = (start + end) / 2;
        int left = start, right = mid + 1;
        
        for (int k = 0; k < n; k++) {
            // merge two sorted arrays
            if ((left <= mid) && (right > end || A[left] <= A[right])) {
                tmp[k] = A[left];
                left++;
            } else {
                tmp[k] = A[right];
                right++;
            }
        }
        
        // assign back to original array A
        for (int k = 0; k < n; k++) {
            A[start + k] = tmp[k];
        }
    }
}

File: algo_basic_python_java/w9.-sorting/problems/464.sort_integers_ii/464_sort_intergers_ii_merge_sort.py
class Solution:
    """
    @param A: an integer array
    @return: nothing
    """
    def sortIntegers2(self, A):
        # write your code here
        if not A: 
            return
        
        tmp = [0 for _ in range(len(A))]
        self.merge_sort(A, 0, len(A) - 1, tmp)
    
    def merge_sort(self, A, start, end, tmp):
        # when split into single element, return 
        if start >= end:
            return 
        mid = start + (end - start)//2
        # keep split in half
        self.merge_sort(A, start, mid, tmp)
        self.merge_sort(A, mid + 1, end, tmp)
        # merge the two half array into tmp
        self.merge(A, start, end, tmp)
    
    def merge(self, A, start, end, tmp):
        left = start
        mid = start + (end - start)//2
        right = mid + 1
        n = end - start + 1
        
        for k in range(n):
            # merge two sorted arrays
            if left <= mid and (right > end or A[left] <= A[right]):
                tmp[k] = A[left]
                left+=1
            else:
                tmp[k] = A[right]
                right+=1
        # assign back to original array A
        for k in range(n):
            A[start + k] = tmp[k]
        

        
   